/*
 * KOTLIN DIAGNOSTICS SPEC TEST (POSITIVE)
 *
 * SECTIONS: constant-literals, real-literals
 * PARAGRAPH: 1
 * SENTENCE: [3] A real literal may also be followed by a type suffix (f or F).
 * NUMBER: 2
 * DESCRIPTION: Real literals suffixed by f/F (the float suffix) with a many digits in a whole-number part and a fraction part  (including conforms).
 */

// TESTCASE NUMBER: 1
val value_1 = 0.000000000000000000000000000000000000000000001f

// TESTCASE NUMBER: 2
val value_2 = <!FLOAT_LITERAL_CONFORMS_ZERO!>0.0000000000000000000000000000000000000000000001f<!>

// TESTCASE NUMBER: 3
val value_3 = -0.00000000000000000000000000000000000000000000123456789012345678901234567890123456789012345678901234567890F

// TESTCASE NUMBER: 4
val value_4 = -<!FLOAT_LITERAL_CONFORMS_ZERO!>0.0000000000000000000000000000000000000000000001f<!>

// TESTCASE NUMBER: 5
val value_5 = -99999999999999999999999999999999999999.0f

// TESTCASE NUMBER: 6
val value_6 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>2222222222222222222222222222222222222222.0f<!>

// TESTCASE NUMBER: 7
val value_7 = -<!FLOAT_LITERAL_CONFORMS_INFINITY!>555555555123456785012345678501234567850.0F<!>

// TESTCASE NUMBER: 8
val value_8 = 333333333333333333333333333333333333333.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001f

// TESTCASE NUMBER: 9
val value_9 = -44444444444444444444444444444444444444.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001F

// TESTCASE NUMBER: 10
val value_10 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>777777777777777777777777777777777777777.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001F<!>

// TESTCASE NUMBER: 11
val value_11 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>888888888888888888888888888888888888888.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000001f<!>

// TESTCASE NUMBER: 12
val value_12 = -<!FLOAT_LITERAL_CONFORMS_INFINITY!>555555555555556785012345678501234567850.0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000123456789012345678901234567890123456789012345678901234567890f<!>
