// DIAGNOSTICS: -CONTEXT_RECEIVERS_DEPRECATED
// WITH_STDLIB
// RENDER_ALL_DIAGNOSTICS_FULL_TEXT
// IGNORE_FIR_DIAGNOSTICS
// IGNORE_ERRORS

<!CONFLICTING_JVM_DECLARATIONS!>object O {
    <!UNSUPPORTED_FEATURE!>context(O)<!>
    @JvmField
    <!CONFLICTING_JVM_DECLARATIONS!>val INSTANCE: O? = null<!>
}<!>
