// DIAGNOSTICS: -CONTEXT_RECEIVERS_DEPRECATED
// WITH_STDLIB
// RENDER_ALL_DIAGNOSTICS_FULL_TEXT
// IGNORE_FIR_DIAGNOSTICS
// IGNORE_ERRORS
// LANGUAGE: +ContextParameters

<!CONFLICTING_JVM_DECLARATIONS!>object O {
    <!CONFLICTING_JVM_DECLARATIONS!><!CONTEXT_PARAMETERS_UNSUPPORTED!>context(_: <!DEBUG_INFO_MISSING_UNRESOLVED!>O<!>)<!>
    @JvmField
    val INSTANCE: O? = null<!>
}<!>
