// WITH_STDLIB
// !RENDER_ALL_DIAGNOSTICS_FULL_TEXT

<!CONFLICTING_JVM_DECLARATIONS!>object O {
    <!CONFLICTING_JVM_DECLARATIONS!><!UNSUPPORTED_FEATURE, UNSUPPORTED_FEATURE!>context(O)<!>
    @JvmField
    val INSTANCE: O? = <!CONTEXT_RECEIVERS_WITH_BACKING_FIELD!>null<!><!>
}<!>