// DIAGNOSTICS: -CONTEXT_RECEIVERS_DEPRECATED
// WITH_STDLIB
// RENDER_ALL_DIAGNOSTICS_FULL_TEXT
// IGNORE_FIR_DIAGNOSTICS
// IGNORE_ERRORS

<!CONFLICTING_JVM_DECLARATIONS!>object O {
    <!CONFLICTING_JVM_DECLARATIONS!><!UNSUPPORTED_FEATURE, UNSUPPORTED_FEATURE!>context(O)<!>
    @JvmField
    val INSTANCE: O? = <!CONTEXT_RECEIVERS_WITH_BACKING_FIELD!>null<!><!>
}<!>