package foo

actual class <!LINE_MARKER("descr='Has declaration in common module'")!>ActualInMiddleCompatibleInBottom<!>
class <!ACTUAL_MISSING!>CompatibleInMiddleActualInBottom<!>

class <!ACTUAL_MISSING!>CompatibleInMiddleAndBottom<!>