val a1: Int = <!INITIALIZER_TYPE_MISMATCH!>1l<!>
val a2: Int = <!INITIALIZER_TYPE_MISMATCH!>0x1l<!>
val a3: Int = <!INITIALIZER_TYPE_MISMATCH!>0X1l<!>
val a4: Int = <!INITIALIZER_TYPE_MISMATCH!>0b1l<!>
val a5: Int = <!INITIALIZER_TYPE_MISMATCH!>0B1l<!>
val a6: Long = 1l
val a7: Long = 0x1l
val a8: Long = 0X1l
val a9: Long = 0b1l
val a10: Long = 0B1l