// RUN_PIPELINE_TILL: FRONTEND
// ISSUE: KT-55281
// CHECK_TYPE_WITH_EXACT

fun test() {
    val buildee = <!CANNOT_INFER_PARAMETER_TYPE!>build<!> {
        this as DerivedBuildee<*>
        consumeNullableAny(getTypeVariable())
    }
    // exact type equality check â€” turns unexpected compile-time behavior into red code
    // considered to be non-user-reproducible code for the purposes of these tests
    checkExactType<<!CANNOT_INFER_PARAMETER_TYPE!>Buildee<Any?><!>>(buildee)
}




class TargetType

fun consumeNullableAny(value: Any?) {}

open class Buildee<TV> {
    fun getTypeVariable(): TV = storage
    private var storage: TV = null!!
}

class DerivedBuildee<TA>: Buildee<TA>()

fun <PTV> build(instructions: Buildee<PTV>.() -> Unit): Buildee<PTV> {
    return DerivedBuildee<PTV>().apply(instructions)
}

/* GENERATED_FIR_TAGS: asExpression, checkNotNullCall, classDeclaration, functionDeclaration, functionalType,
intersectionType, lambdaLiteral, localProperty, nullableType, propertyDeclaration, smartcast, starProjection,
stringLiteral, thisExpression, typeParameter, typeWithExtension */
