// IGNORE_FIR_DIAGNOSTICS
// RUN_PIPELINE_TILL: FIR2IR
// DISABLE_NEXT_PHASE_SUGGESTION: we need to run fi2ir to get all actualization diagnostics
// LATEST_LV_DIFFERENCE

// MODULE: common

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}, PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>C<!>

// MODULE: intermediate()()(common)

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION, PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>B<!>

// MODULE: main()()(common, intermediate)

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>B<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>C<!>
