// MODULE: common
// TARGET_PLATFORM: Common

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{COMMON}, PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>C<!>

// MODULE: intermediate()()(common)
// TARGET_PLATFORM: Common

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION, PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>B<!>

// MODULE: main()()(common, intermediate)

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>B<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>C<!>
