// IGNORE_FIR_DIAGNOSTICS
// RUN_PIPELINE_TILL: FRONTEND
// MODULE: common

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}, PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>C<!>

// MODULE: intermediate()()(common)

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION, PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>B<!>

// MODULE: main()()(common, intermediate)

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>B<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>C<!>
