// IGNORE_FIR_DIAGNOSTICS
// RUN_PIPELINE_TILL: BACKEND
// LATEST_LV_DIFFERENCE

// MODULE: common

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM;JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>C<!>

// MODULE: intermediate()()(common)

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION, PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>A<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION{JVM}!>B<!>

// MODULE: main()()(common, intermediate)

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>B<!>

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>C<!>

/* GENERATED_FIR_TAGS: classDeclaration */
