// RUN_PIPELINE_TILL: FRONTEND
// FIR_IDENTICAL
// RENDER_DIAGNOSTICS_FULL_TEXT

class A1<T0, F0, E0> where T0 : F0?, F0 : <!CYCLIC_GENERIC_UPPER_BOUND!>E0<!>, E0 : <!CYCLIC_GENERIC_UPPER_BOUND!>F0<!>

class A2<T1, F1, E1, K1> where T1 : F1?, F1 : <!CYCLIC_GENERIC_UPPER_BOUND!>E1<!>, E1 : <!CYCLIC_GENERIC_UPPER_BOUND!>F1<!>, E1 : <!CYCLIC_GENERIC_UPPER_BOUND!>K1<!>, K1 : <!CYCLIC_GENERIC_UPPER_BOUND!>E1<!>

/* GENERATED_FIR_TAGS: classDeclaration, nullableType, typeConstraint, typeParameter */
