// WITH_STDLIB

fun foo() {
    <!INITIALIZER_REQUIRED_FOR_DESTRUCTURING_DECLARATION!>val (a, b) by listOf(1, 2)<!>
    <!INITIALIZER_REQUIRED_FOR_DESTRUCTURING_DECLARATION!>val (c, d) by lazy { listOf(3, 4) }<!>
}
