// !CHECK_TYPE

/*
 * KOTLIN DIAGNOSTICS SPEC TEST (NEGATIVE)
 *
 * SPEC VERSION: 0.1-draft
 * PLACE: constant-literals, boolean-literals -> paragraph 1 -> sentence 3
 * NUMBER: 1
 * DESCRIPTION: The use of Boolean literals as the identifier (with backtick) in the class.
 * NOTE: this test data is generated by FeatureInteractionTestDataGenerator. DO NOT MODIFY CODE MANUALLY
 */

// TESTCASE NUMBER: 1
fun case_1() {
    true checkType { <!TYPE_MISMATCH!>_<!><Boolean?>() }
    false checkType { <!TYPE_MISMATCH!>_<!><Boolean?>() }

    true checkType { <!TYPE_MISMATCH!>_<!><Any?>() }
    false checkType { <!TYPE_MISMATCH!>_<!><Any>() }

    true checkType { <!TYPE_MISMATCH!>_<!><Nothing?>() }
    false checkType { <!TYPE_MISMATCH!>_<!><Nothing>() }
}