// !LANGUAGE: +NewInference
// !DIAGNOSTICS: -UNUSED_VARIABLE -ASSIGNED_BUT_NEVER_ACCESSED_VARIABLE -UNUSED_VALUE -UNUSED_PARAMETER -UNUSED_EXPRESSION
// SKIP_TXT

// TESTCASE NUMBER: 1
fun case1() {
    val c = <!EMPTY_CHARACTER_LITERAL!>''<!>
}

// TESTCASE NUMBER: 2

fun case2() {
    val c2: Char = <!EMPTY_CHARACTER_LITERAL!>''<!><!SYNTAX!>'<!>
    val c3: Char = <!ILLEGAL_ESCAPE!>'\'<!>
}

// TESTCASE NUMBER: 3

fun case3() {
    val c1: Char = <!TOO_MANY_CHARACTERS_IN_CHARACTER_LITERAL!>'B a'<!>
    val c2: Char = <!TOO_MANY_CHARACTERS_IN_CHARACTER_LITERAL!>'  '<!>
    val c3: Char = <!TOO_MANY_CHARACTERS_IN_CHARACTER_LITERAL!>'Ba'<!>
}

// TESTCASE NUMBER: 4

fun case4() {
    val cOutOfRaneMin = <!TOO_MANY_CHARACTERS_IN_CHARACTER_LITERAL!>'êÄÄ'<!> //u+10000

    val cOutOfRangeAroundMax = <!TOO_MANY_CHARACTERS_IN_CHARACTER_LITERAL!>'Û†áøÛ†áøÛüøø'<!> //u+Dfffff
}
