fun <!VIPER_TEXT!>return_unit<!>() {}
fun <!VIPER_TEXT!>return_int<!>(): Int { return 0 }
fun <!VIPER_TEXT!>take_int_return_unit<!>(@Suppress("UNUSED_PARAMETER") x: Int) {}
fun <!VIPER_TEXT!>take_int_return_int<!>(x: Int): Int { return x }
