class A<T1, <!REIFIED_TYPE_PARAMETER_NO_INLINE!>reified<!> T2, T3, <!REIFIED_TYPE_PARAMETER_NO_INLINE!>reified<!> T4> {
    fun<<!REIFIED_TYPE_PARAMETER_NO_INLINE!>reified<!> R> foo(): T2 = throw UnsupportedOperationException()
}
