// RUN_PIPELINE_TILL: FRONTEND
// ISSUE: KT-82638
// LANGUAGE: +CollectionLiterals
// WITH_STDLIB

fun foo(a: LongArray) {
}

fun foo(b: IntArray) {
}

fun test() {
    <!OVERLOAD_RESOLUTION_AMBIGUITY!>foo<!>(<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[]<!>)
    <!OVERLOAD_RESOLUTION_AMBIGUITY!>foo<!>(<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[42]<!>)
    <!OVERLOAD_RESOLUTION_AMBIGUITY!>foo<!>(<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>["42"]<!>)
    <!OVERLOAD_RESOLUTION_AMBIGUITY!>foo<!>(<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[42u]<!>)
    <!OVERLOAD_RESOLUTION_AMBIGUITY!>foo<!>(<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[42L]<!>)
    <!OVERLOAD_RESOLUTION_AMBIGUITY!>foo<!>(<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[42, 42L]<!>)
}

/* GENERATED_FIR_TAGS: collectionLiteral, functionDeclaration, integerLiteral, stringLiteral, unsignedLiteral */
