fun <!VIPER_TEXT!>less<!>(x: Int, y: Int): Boolean {
    return x < y
}

fun <!VIPER_TEXT!>less_equal<!>(x: Int, y: Int): Boolean {
    return x <= y
}

fun <!VIPER_TEXT!>greater<!>(x: Int, y: Int): Boolean {
    return x > y
}

fun <!VIPER_TEXT!>greater_equal<!>(x: Int, y: Int): Boolean {
    return x >= y
}
