fun <!VIPER_TEXT!>simple_if<!>(): Int {
    if (true) {
        return 0
    } else {
        return 1
    }
}
fun <!VIPER_TEXT!>if_on_parameter<!>(b: Boolean): Int {
    if (b) {
        return 0
    } else {
        return 1
    }
}
