fun foo() {
    <!UNRESOLVED_REFERENCE!>x<!>.<!DEBUG_INFO_MISSING_UNRESOLVED!>yyy<!><<!UNRESOLVED_REFERENCE!>XXX<!>>()
    <!UNRESOLVED_REFERENCE!>x<!>.<!DEBUG_INFO_MISSING_UNRESOLVED!>yyy<!><Int>()
}