// !WITH_NEW_INFERENCE
fun foo(f: String.() -> Int) {}
val test = foo(<!TYPE_MISMATCH, TYPE_MISMATCH, TYPE_MISMATCH!>fun <!EXPECTED_PARAMETERS_NUMBER_MISMATCH{NI}!>()<!> = <!UNRESOLVED_REFERENCE!>length<!><!>)
