// JAVAC_SKIP
typealias Test1 = <!UNRESOLVED_REFERENCE!>SuspendFunction0<Unit><!>
typealias Test2 = <!UNRESOLVED_REFERENCE!>kotlin.SuspendFunction0<Unit><!>
typealias Test3 = kotlin.coroutines.SuspendFunction0<Unit>