// RUN_PIPELINE_TILL: FRONTEND
fun main() {
    <!NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER, UNSUPPORTED!>[]<!>!!<!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!>
}

/* GENERATED_FIR_TAGS: checkNotNullCall, collectionLiteral, functionDeclaration */
