// RUN_PIPELINE_TILL: FRONTEND
fun main() {
    <!UNSUPPORTED_ARRAY_LITERAL_OUTSIDE_OF_ANNOTATION_ERROR!>[]<!><!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!><!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!>
}

/* GENERATED_FIR_TAGS: checkNotNullCall, collectionLiteral, functionDeclaration */
