// RUN_PIPELINE_TILL: FRONTEND
fun main() {
    <!UNSUPPORTED!>[]<!><!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!><!UNNECESSARY_NOT_NULL_ASSERTION!>!!<!>
}

/* GENERATED_FIR_TAGS: checkNotNullCall, collectionLiteral, functionDeclaration */
