// !DIAGNOSTICS: -UNUSED_PARAMETER
open class B(x: Int)
<!SUPERTYPE_INITIALIZED_WITHOUT_PRIMARY_CONSTRUCTOR!>class <!CONFLICTING_OVERLOADS!>A<!> : B(1) {
    <!CONFLICTING_OVERLOADS!>constructor()<!>: super(1)
}<!>
