// FIR_IDENTICAL
class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>A<!>
class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>A<!> {
    constructor()
}

class B
class Outer {
    class B {
        constructor()
    }
}
