// RUN_PIPELINE_TILL: FRONTEND
inline operator fun <reified T> Int.invoke() = this

val a2 = <!CANNOT_INFER_PARAMETER_TYPE, NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER!>1()<!>
val a3 = 1.<!CANNOT_INFER_PARAMETER_TYPE, NEW_INFERENCE_NO_INFORMATION_FOR_PARAMETER!>invoke<!>()
