// RUN_PIPELINE_TILL: FRONTEND
<!MUST_BE_INITIALIZED!>val <<!INCORRECT_TYPE_PARAMETER_OF_PROPERTY!>T : <!UNRESOLVED_REFERENCE!>KClass<!><T>.something<!>> abc<!>

/* GENERATED_FIR_TAGS: propertyDeclaration, typeConstraint, typeParameter */
