fun test() {
    <!INVISIBLE_REFERENCE!>Double<!>()
    <!INVISIBLE_REFERENCE!>Float<!>()
    <!INVISIBLE_REFERENCE!>Long<!>()
    <!INVISIBLE_REFERENCE!>Int<!>()
    <!INVISIBLE_REFERENCE!>Short<!>()
    <!INVISIBLE_REFERENCE!>Byte<!>()
    <!INVISIBLE_REFERENCE!>Char<!>()
    <!INVISIBLE_REFERENCE!>Boolean<!>()
}
