fun test() {
    <!INVISIBLE_MEMBER!>Double<!>()
    <!INVISIBLE_MEMBER!>Float<!>()
    <!INVISIBLE_MEMBER!>Long<!>()
    <!INVISIBLE_MEMBER!>Int<!>()
    <!INVISIBLE_MEMBER!>Short<!>()
    <!INVISIBLE_MEMBER!>Byte<!>()
    <!INVISIBLE_MEMBER!>Char<!>()
    <!INVISIBLE_MEMBER!>Boolean<!>()
}