fun test() {
    Double(<!NO_VALUE_FOR_PARAMETER!>)<!>
    Float(<!NO_VALUE_FOR_PARAMETER!>)<!>
    Long(<!NO_VALUE_FOR_PARAMETER!>)<!>
    Int(<!NO_VALUE_FOR_PARAMETER!>)<!>
    Short(<!NO_VALUE_FOR_PARAMETER!>)<!>
    Byte(<!NO_VALUE_FOR_PARAMETER!>)<!>
    <!INVISIBLE_REFERENCE!>Char<!>()
    Boolean(<!NO_VALUE_FOR_PARAMETER!>)<!>
}
