fun test() {
    <!NO_VALUE_FOR_PARAMETER!>Double()<!>
    <!NO_VALUE_FOR_PARAMETER!>Float()<!>
    <!NO_VALUE_FOR_PARAMETER!>Long()<!>
    <!NO_VALUE_FOR_PARAMETER!>Int()<!>
    <!NO_VALUE_FOR_PARAMETER!>Short()<!>
    <!NO_VALUE_FOR_PARAMETER!>Byte()<!>
    <!INVISIBLE_REFERENCE!>Char<!>()
    <!NO_VALUE_FOR_PARAMETER!>Boolean()<!>
}
