// RUN_PIPELINE_TILL: BACKEND
// ISSUE: KT-73508
// JDK_KIND: FULL_JDK_21
// DIAGNOSTICS: -ASSIGNED_BUT_NEVER_ACCESSED_VARIAB

// FILE: J.kt

import java.util.concurrent.atomic.AtomicReference
import java.util.concurrent.atomic.AtomicReferenceArray

fun testReference() {
    val it = AtomicReference<Int>(127)
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndSet(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.<!DEPRECATION!>weakCompareAndSet<!>(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetAcquire(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetRelease(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetPlain(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetVolatile(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndExchange(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndExchangeAcquire(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndExchangeRelease(<!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
}

fun testReferenceArray() {
    val it = AtomicReferenceArray<Int>(1)
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndSet(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.<!DEPRECATION!>weakCompareAndSet<!>(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetAcquire(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetRelease(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetPlain(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.weakCompareAndSetVolatile(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndExchange(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndExchangeAcquire(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
    <!ATOMIC_REF_WITHOUT_CONSISTENT_IDENTITY!>it.compareAndExchangeRelease(0, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>127<!>, <!ATOMIC_REF_CALL_ARGUMENT_WITHOUT_CONSISTENT_IDENTITY!>128<!>)<!>
}

/* GENERATED_FIR_TAGS: flexibleType, functionDeclaration, integerLiteral, javaFunction, localProperty,
propertyDeclaration */
