// FIR_IDENTICAL

fun main() {
    val appendable = <!NO_COMPANION_OBJECT!>Appendable<!>
    val stringBuilder = <!NO_COMPANION_OBJECT!>StringBuilder<!>
    val exception = <!NO_COMPANION_OBJECT!>IllegalStateException<!>
}
