interface MyTrait: <!PLATFORM_CLASS_MAPPED_TO_KOTLIN, TRAIT_WITH_SUPERCLASS!>Object<!> {
    override fun toString(): String
    public override fun finalize()
    public <!OVERRIDING_FINAL_MEMBER!>override<!> fun wait()
}