// RUN_PIPELINE_TILL: FRONTEND
// DIAGNOSTICS: -NO_VALUE_FOR_PARAMETER
class Tree<T>(T <!SYNTAX!>element<!>, <!SYNTAX!><!>Tree<T><!SYNTAX!><!> left<!SYNTAX!><!>, <!SYNTAX!><!>Tree<T><!SYNTAX!><!> right<!SYNTAX!><!>) {}