val <!ELEMENT!> = 10
val <!ELEMENT!> = "."

val value_1 = <!ELEMENT!> - 100 % <!ELEMENT!>
val value_2 = <!ELEMENT!>.dec()
val value_3 = "$<!ELEMENT!> 10"
val value_4 = "${<!ELEMENT!>}"
val value_5 = <!ELEMENT!> + " 11..." + <!ELEMENT!> + "1"
val value_6 = <!ELEMENT!>
