/*
 * KOTLIN DIAGNOSTICS SPEC TEST (NEGATIVE)
 *
 * SPEC VERSION: 0.1-100
 * PLACE: expressions, constant-literals, real-literals -> paragraph 4 -> sentence 1
 * NUMBER: 2
 * DESCRIPTION: Real literals with underscores at the end.
 */

// TESTCASE NUMBER: 1
val value_1= <!ILLEGAL_UNDERSCORE!>.0_0e-0___<!>

// TESTCASE NUMBER: 2
val value_2 = <!ILLEGAL_UNDERSCORE!>0_0___0E+10_<!>

// TESTCASE NUMBER: 3
val value_3 = <!ILLEGAL_UNDERSCORE!>.0_0e-0_F<!>

// TESTCASE NUMBER: 4
val value_4 = <!ILLEGAL_UNDERSCORE!>0__________________________0E+3___<!>

// TESTCASE NUMBER: 5
val value_5 = <!ILLEGAL_UNDERSCORE!>0_00e1_<!>

// TESTCASE NUMBER: 6
val value_6 = <!ILLEGAL_UNDERSCORE!>.0_0__0___0E0_______________<!>

// TESTCASE NUMBER: 7
val value_7 = <!FLOAT_LITERAL_CONFORMS_INFINITY, ILLEGAL_UNDERSCORE!>0_1e+100_____f<!>

// TESTCASE NUMBER: 8
val value_8 = <!ILLEGAL_UNDERSCORE!>.0_0E-0001___<!>

// TESTCASE NUMBER: 9
val value_9 = <!ILLEGAL_UNDERSCORE!>0___0__0e4___<!>

// TESTCASE NUMBER: 10
val value_10 = <!ILLEGAL_UNDERSCORE!>0_00_0e99_f<!>

// TESTCASE NUMBER: 11
val value_11 = <!FLOAT_LITERAL_CONFORMS_ZERO, ILLEGAL_UNDERSCORE!>.0_5E-93_______F<!>

// TESTCASE NUMBER: 12
val value_12 = <!ILLEGAL_UNDERSCORE!>.8____8888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888_____8888888888888888888e+5_f<!>

// TESTCASE NUMBER: 13
val value_13 = <!ILLEGAL_UNDERSCORE!>0_0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000_0e40_F<!>

// TESTCASE NUMBER: 14
val value_14 = <!ILLEGAL_UNDERSCORE!>0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_0_e-10_f<!>

// TESTCASE NUMBER: 15
val value_15 = <!ILLEGAL_UNDERSCORE!>.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e+75_<!>
