/*
 * KOTLIN DIAGNOSTICS SPEC TEST (NEGATIVE)
 *
 * SPEC VERSION: 0.1-100
 * MAIN LINK: expressions, constant-literals, real-literals -> paragraph 4 -> sentence 2
 * NUMBER: 2
 * DESCRIPTION: Real literals with underscores after an exponent mark.
 */

// TESTCASE NUMBER: 1
val value_1 = <!ILLEGAL_UNDERSCORE!>0e___0<!>

// TESTCASE NUMBER: 2
val value_2 = <!ILLEGAL_UNDERSCORE!>.0e_0<!>

// TESTCASE NUMBER: 3
val value_3 = <!ILLEGAL_UNDERSCORE!>0.1e__2<!>

// TESTCASE NUMBER: 4
val value_4 = <!ILLEGAL_UNDERSCORE!>2.1e_2<!>

// TESTCASE NUMBER: 5
val value_5 = <!ILLEGAL_UNDERSCORE!>1E-_2F<!>

// TESTCASE NUMBER: 6
val value_6 = <!ILLEGAL_UNDERSCORE!>3e+_0f<!>

// TESTCASE NUMBER: 7
val value_7 = <!ILLEGAL_UNDERSCORE!>5E_<!><!OVERLOAD_RESOLUTION_AMBIGUITY!>-<!>0

// TESTCASE NUMBER: 8
val value_8 = <!ILLEGAL_UNDERSCORE!>5e_<!><!OVERLOAD_RESOLUTION_AMBIGUITY!>-<!>0F

// TESTCASE NUMBER: 9
val value_9 = <!ILLEGAL_UNDERSCORE!>5e_00000000<!>

// TESTCASE NUMBER: 10
val value_10 = <!ILLEGAL_UNDERSCORE!>0.0_0e_1_0f<!>

// TESTCASE NUMBER: 11
val value_11 = <!ILLEGAL_UNDERSCORE!>0.0_0E____<!><!OVERLOAD_RESOLUTION_AMBIGUITY!>-<!>0__0_0F

// TESTCASE NUMBER: 12
val value_12 = <!ILLEGAL_UNDERSCORE!>.0_0E__0_0<!>

// TESTCASE NUMBER: 13
val value_13 = <!ILLEGAL_UNDERSCORE!>00_______________00.0_0e+__0_0<!>

// TESTCASE NUMBER: 14
val value_14 = <!ILLEGAL_UNDERSCORE!>.0e-__1_0F<!>

// TESTCASE NUMBER: 15
val value_15 = <!ILLEGAL_UNDERSCORE!>33__3.0e_10__0<!>

// TESTCASE NUMBER: 16
val value_16 = <!ILLEGAL_UNDERSCORE!>.0E_0______00F<!>

// TESTCASE NUMBER: 17
val value_17 = <!ILLEGAL_UNDERSCORE!>666_666.0__________________________________________________1E+_2___________________________________________________________________0F<!>

// TESTCASE NUMBER: 18
val value_18 = <!ILLEGAL_UNDERSCORE!>8888888_8.000e_____0f<!>

// TESTCASE NUMBER: 19
val value_19 = <!ILLEGAL_UNDERSCORE!>9_______9______9_____9____9___9__9_9.0E__<!><!OVERLOAD_RESOLUTION_AMBIGUITY!>-<!>1

// TESTCASE NUMBER: 20
val value_20 = <!ILLEGAL_UNDERSCORE!>0_0_0_0_0_0_0_0_0_0.12345678e+__90F<!>

// TESTCASE NUMBER: 21
val value_21 = <!ILLEGAL_UNDERSCORE!>1_2_3_4_5_6_7_8_9.2_3_4_5_6_7_8_9e_<!>-<!UNRESOLVED_REFERENCE!>_0<!>

// TESTCASE NUMBER: 22
val value_22 = <!ILLEGAL_UNDERSCORE!>.345______________6e_____7_______8f<!>

// TESTCASE NUMBER: 23
val value_23 = <!ILLEGAL_UNDERSCORE!>.45_6E_7f<!>

// TESTCASE NUMBER: 24
val value_24 = <!ILLEGAL_UNDERSCORE!>6_54.765e-_4<!>

// TESTCASE NUMBER: 25
val value_25 = <!ILLEGAL_UNDERSCORE!>7_6543.8E____7654_3<!>

// TESTCASE NUMBER: 26
val value_26 = <!ILLEGAL_UNDERSCORE!>.9E+_______0_____________8765432f<!>

// TESTCASE NUMBER: 27
val value_27 = <!ILLEGAL_UNDERSCORE!>000000000000000000000000000000000000000000000000000000000000000000000000000000000000000___0.000000000000000000000000e_000000000000000000000000000000000000000000000000000000000000000_0F<!>

// TESTCASE NUMBER: 28
val value_28 = <!ILLEGAL_UNDERSCORE!>0_000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.0E-_0___000000000000000000000000000000000000000000000000000000000000000000000000000000000000000<!>

// TESTCASE NUMBER: 29
val value_29 = <!ILLEGAL_UNDERSCORE!>9999999999999999999999999999999999999999999_______________999999999999999999999999999999999999999999999.33333333333333333333333333333333333333333333333_333333333333333333333333333333333333333e___3_3f<!>
