const val p1: Int = <!CONST_VAL_WITH_NON_CONST_INITIALIZER!>'\n'.code<!>
const val p2: Long = <!CONST_VAL_WITH_NON_CONST_INITIALIZER!>'\n'.code.toLong()<!>
const val p3: Byte = <!CONST_VAL_WITH_NON_CONST_INITIALIZER!>'\n'.code.toByte()<!>
const val p4: Short = <!CONST_VAL_WITH_NON_CONST_INITIALIZER!>'\n'.code.toShort()<!>

const val e2: Long = <!CONST_VAL_WITH_NON_CONST_INITIALIZER!>'\n'.code<!>
const val e3: Byte = <!CONST_VAL_WITH_NON_CONST_INITIALIZER!>'\n'.code<!>
const val e4: Short = <!CONST_VAL_WITH_NON_CONST_INITIALIZER!>'\n'.code<!>
