// NI_EXPECTED_FILE

val unwrapped = <!UNRESOLVED_REFERENCE!>some<!>.<!SYNTAX!><<!>cabc<!SYNTAX!><!SYNTAX!>$Wrapper<!><<!PROJECTION_ON_NON_CLASS_TYPE_ARGUMENT!>out<!> Any><!>::<!UNRESOLVED_REFERENCE!>unwrap<!>
