/*
 * KOTLIN DIAGNOSTICS SPEC TEST (POSITIVE)
 *
 * SECTIONS: constant-literals, real-literals
 * PARAGRAPH: 2
 * SENTENCE: [1] The exponent is an exponent mark (e or E) followed by an optionaly signed decimal integer (a sequence of decimal digits).
 * NUMBER: 2
 * DESCRIPTION: Real literals suffixed by f/F (float suffix) with an exponent mark.
 */

// TESTCASE NUMBER: 1
val value_1 = 0.0e0f

// TESTCASE NUMBER: 2
val value_2 = 0.0e-00F

// TESTCASE NUMBER: 3
val value_3 = 0.0E000F

// TESTCASE NUMBER: 4
val value_4 = 0.0E+0000f

// TESTCASE NUMBER: 5
val value_5 = 00.0e+0f

// TESTCASE NUMBER: 6
val value_6 = 000.00e00f

// TESTCASE NUMBER: 7
val value_7 = 0000.000E-000F

// TESTCASE NUMBER: 8
val value_8 = 1.0E+1F

// TESTCASE NUMBER: 9
val value_9 = 22.00e22F

// TESTCASE NUMBER: 10
val value_10 = 2345678.345678e00000000001F

// TESTCASE NUMBER: 11
val value_11 = 456.56e-0f

// TESTCASE NUMBER: 12
val value_12 = 5.65e000000000000F

// TESTCASE NUMBER: 13
val value_13 = 654.7654E+010f

// TESTCASE NUMBER: 14
val value_14 = 76543.876543E1f

// TESTCASE NUMBER: 15
val value_15 = 8765432.98765432e-2f

// TESTCASE NUMBER: 16
val value_16 = 987654321.0987654321E-3f

// TESTCASE NUMBER: 17
val value_17 = 0.1111e4f

// TESTCASE NUMBER: 18
val value_18 = 1.22222E-5F

// TESTCASE NUMBER: 19
val value_19 = 9.33333e+6F

// TESTCASE NUMBER: 20
val value_20 = 9.444444E7F

// TESTCASE NUMBER: 21
val value_21 = 8.5555555e8f

// TESTCASE NUMBER: 22
val value_22 = <!FLOAT_LITERAL_CONFORMS_ZERO!>3.777777777E-308f<!>

// TESTCASE NUMBER: 23
val value_23 = <!FLOAT_LITERAL_CONFORMS_ZERO!>6.99999999999e-309F<!>

// TESTCASE NUMBER: 24
val value_24 = 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f

// TESTCASE NUMBER: 25
val value_25 = 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e-000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000F

// TESTCASE NUMBER: 26
val value_26 = 0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e+000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000f
