// RUN_PIPELINE_TILL: FRONTEND
// WITH_STDLIB
// ISSUE: KT-83241

val <!SYNTAX!>(a, b)<!> = run <!ARGUMENT_TYPE_MISMATCH!>{ <!CANNOT_INFER_IT_PARAMETER_TYPE, COMPONENT_FUNCTION_AMBIGUITY, COMPONENT_FUNCTION_AMBIGUITY!>(c, d)<!> -> }<!>

/* GENERATED_FIR_TAGS: lambdaLiteral, localProperty, propertyDeclaration */
