// IGNORE_FIR_DIAGNOSTICS
// RUN_PIPELINE_TILL: FIR2IR
// DIAGNOSTICS: -UNUSED_PARAMETER
// MODULE: m1-common
// FILE: common.kt

<!EXPECT_ACTUAL_IR_INCOMPATIBILITY{JVM;JVM}!>expect<!> class Foo1
<!EXPECT_ACTUAL_IR_INCOMPATIBILITY{JVM;JVM}!>expect<!> class Foo2

expect fun foo2(): Int

<!EXPECT_ACTUAL_IR_INCOMPATIBILITY{JVM}!>expect<!> val s: String

<!EXPECT_ACTUAL_IR_INCOMPATIBILITY{JVM}!>expect<!> open class Foo3

// MODULE: m2-jvm()()(m1-common)

// FILE: jvm.kt

interface <!ACTUAL_MISSING!>Foo1<!>
actual interface <!EXPECT_ACTUAL_INCOMPATIBILITY_CLASS_KIND, EXPECT_ACTUAL_INCOMPATIBILITY_MODALITY!>Foo2<!>

actual var <!EXPECT_ACTUAL_INCOMPATIBILITY_PROPERTY_KIND!>s<!>: String = "value"

fun <!ACTUAL_MISSING!>foo2<!>(): Int = 0

actual class <!CLASSIFIER_REDECLARATION, EXPECT_ACTUAL_INCOMPATIBILITY_MODALITY!>Foo3<!>

class <!ACTUAL_MISSING, CLASSIFIER_REDECLARATION!>Foo3<!>
