// NEVER_VALIDATE

fun <!VIPER_TEXT!>less<!>(x: Int, y: Int): Boolean {
    return x < y
}

fun <!VIPER_TEXT!>lessEqual<!>(x: Int, y: Int): Boolean {
    return x <= y
}

fun <!VIPER_TEXT!>greater<!>(x: Int, y: Int): Boolean {
    return x > y
}

fun <!VIPER_TEXT!>greaterEqual<!>(x: Int, y: Int): Boolean {
    return x >= y
}
