fun box()
{
    var <!UNUSED_VARIABLE!>b<!> = #(0)
    var <!UNUSED_VARIABLE!>b1<!> = #(1)
    var <!UNUSED_VARIABLE!>b2<!> = #(1,1)
    var <!UNUSED_VARIABLE!>b3<!> = #(1,1,1)
    var <!UNUSED_VARIABLE!>b4<!> = #(1,1,1,1)
    var <!UNUSED_VARIABLE!>b5<!> = #(1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b6<!> = #(1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b7<!> = #(1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b8<!> = #(1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b9<!> = #(1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b10<!> = #(1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b11<!> = #(1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b12<!> = #(1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b13<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b14<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b15<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b16<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b17<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b18<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b19<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b20<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b21<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b22<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
    var <!UNUSED_VARIABLE!>b23<!> = #(1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1)
}