// RUN_PIPELINE_TILL: FRONTEND
// FIR_IDENTICAL

@Deprecated(message = "", level = DeprecationLevel.HIDDEN)
class <!CONFLICTING_OVERLOADS!>KlassA<!>
<!CONFLICTING_OVERLOADS!>fun KlassA()<!> {}

@Deprecated(message = "", level = DeprecationLevel.HIDDEN)
class KlassB<!CONFLICTING_OVERLOADS!>()<!>
<!CONFLICTING_OVERLOADS!>fun KlassB()<!> {}

@Deprecated(message = "", level = DeprecationLevel.HIDDEN)
class KlassC <!CONFLICTING_OVERLOADS!>constructor()<!>
<!CONFLICTING_OVERLOADS!>fun KlassC()<!> {}

@Deprecated(message = "", level = DeprecationLevel.HIDDEN)
class KlassD {
    <!CONFLICTING_OVERLOADS!>constructor()<!>
}
<!CONFLICTING_OVERLOADS!>fun KlassD()<!> {}
