var err: Int
    set(value<!SYNTAX, SYNTAX!><!> = 42<!SYNTAX!>)<!> <!FUNCTION_DECLARATION_WITH_NO_NAME!><!SYNTAX!><!>{
        <!UNRESOLVED_REFERENCE!>field<!> = <!UNRESOLVED_REFERENCE!>value<!>
    }<!>
