fun foo() {
    <!WRONG_MODIFIER_TARGET!>public<!> class A
    <!WRONG_MODIFIER_TARGET!>private<!> class B
    <!WRONG_MODIFIER_TARGET!>protected<!> class C
    <!WRONG_MODIFIER_TARGET!>internal<!> class D
}