class A {
    fun bar() {
        <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
            <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                    <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                        <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                            <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                    <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                        <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                            <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                    <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                        <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                            <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                    <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                        <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                            <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                                <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                                    <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                                        <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                                            <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                                                <!UNRESOLVED_REFERENCE!>div<!> <!UNRESOLVED_REFERENCE!>{
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                    <!UNRESOLVED_REFERENCE!>+<!>(<!UNRESOLVED_REFERENCE!>foo<!> ?: "")
                                                                                                }<!>
                                                                                            }<!>
                                                                                        }<!>
                                                                                    }<!>
                                                                                }<!>
                                                                            }<!>
                                                                        }<!>
                                                                    }<!>
                                                                }<!>
                                                            }<!>
                                                        }<!>

                                                    }<!>
                                                }<!>
                                            }<!>
                                        }<!>
                                    }<!>
                                }<!>
                            }<!>
                        }<!>
                    }<!>
                }<!>
            }<!>
        }<!>
    }
}
