// API_VERSION: 1.7

import kotlin.io.path.<!API_NOT_AVAILABLE!>OnErrorResult<!>

fun fun0 (): Unit {
    val something1 = <!API_NOT_AVAILABLE, OPT_IN_USAGE_ERROR!>OnErrorResult<!>.<!OPT_IN_USAGE_ERROR!>TERMINATE<!>
    val something2 = kotlin.io.path.<!API_NOT_AVAILABLE, OPT_IN_USAGE_ERROR!>OnErrorResult<!>.<!OPT_IN_USAGE_ERROR!>TERMINATE<!>
}
