// FILE: a.kt
@file:JvmName("Util")
@file:JvmMultifileClass

<!CONFLICTING_JVM_DECLARATIONS, CONFLICTING_JVM_DECLARATIONS!>fun main()<!> {}

// FILE: b.kt
@file:JvmName("Util")
@file:JvmMultifileClass

<!CONFLICTING_JVM_DECLARATIONS, CONFLICTING_JVM_DECLARATIONS!>fun main()<!> {}