open class Required

interface Trait : <!TRAIT_WITH_SUPERCLASS!>Required<!>

abstract <!UNMET_TRAIT_REQUIREMENT!>class Abstract<!> : Trait

abstract <!UNMET_TRAIT_REQUIREMENT!>class AbstractDerived<!> : Abstract()