open class Generic<T>

trait A : <!TRAIT_WITH_SUPERCLASS!>Generic<Int><!>

trait B : <!TRAIT_WITH_SUPERCLASS!>Generic<String><!>

<!UNMET_TRAIT_REQUIREMENT, UNMET_TRAIT_REQUIREMENT!>class Y<!> : <!INCONSISTENT_TYPE_PARAMETER_VALUES!>A, B<!>
class Z : <!INCONSISTENT_TYPE_PARAMETER_VALUES!>A, B, Generic<Int>()<!>
