open class Required

trait A : <!TRAIT_WITH_SUPERCLASS!>Required<!>

trait B : A, <!TRAIT_WITH_SUPERCLASS!>Required<!>

trait C : <!TRAIT_WITH_SUPERCLASS!>Required<!>

trait D : B, <!TRAIT_WITH_SUPERCLASS!>Required<!>

<!UNMET_TRAIT_REQUIREMENT, UNMET_TRAIT_REQUIREMENT, UNMET_TRAIT_REQUIREMENT!>class W<!> : D
class X : D, Required()
<!UNMET_TRAIT_REQUIREMENT, UNMET_TRAIT_REQUIREMENT, UNMET_TRAIT_REQUIREMENT, UNMET_TRAIT_REQUIREMENT!>class Y<!> : C, D
class Z : D, C, Required()
