// TESTCASE NUMBER: 1
val value_1 = <!ILLEGAL_UNDERSCORE!>0x3_4_5_6_7_8_____<!>

// TESTCASE NUMBER: 2
val value_2 = <!ILLEGAL_UNDERSCORE!>0X4_______5_______6_______7_<!>

// TESTCASE NUMBER: 3
val value_3 = <!ILLEGAL_UNDERSCORE!>0X000000000_<!>

// TESTCASE NUMBER: 5
val value_5 = <!ILLEGAL_UNDERSCORE!>0x_<!>

// TESTCASE NUMBER: 6
val value_6 = <!ILLEGAL_UNDERSCORE!>0X______________<!>

// TESTCASE NUMBER: 7
val value_7 = <!ILLEGAL_UNDERSCORE!>0X0_<!>

// TESTCASE NUMBER: 8
val value_8 = <!ILLEGAL_UNDERSCORE!>0X10_<!>
