// !DIAGNOSTICS: -UNUSED_PARAMETER
<!ILLEGAL_JVM_NAME!>@jvmName("")<!>
fun foo(a: Any) {}

<!ILLEGAL_JVM_NAME!>@jvmName(".")<!>
fun foo() {}

<!ILLEGAL_JVM_NAME!>@jvmName("/")<!>
fun fooSlash() {}

<!ILLEGAL_JVM_NAME!>@jvmName("<")<!>
fun fooLT() {}