<!UNRESOLVED_REFERENCE!>@Ann<!> class A
<!UNRESOLVED_REFERENCE!>@Ann<!> class B
<!UNRESOLVED_REFERENCE!>@Ann<!>(1) class C
@kotlin.<!UNRESOLVED_REFERENCE!>Ann<!>(1) class D
@kotlin.annotation.<!UNRESOLVED_REFERENCE!>Ann<!>(1) class E