// TESTCASE NUMBER: 1
val value_1 = <!ILLEGAL_CONST_EXPRESSION!>.0E<!>

// TESTCASE NUMBER: 2
val value_2 = <!ILLEGAL_CONST_EXPRESSION!>.0e<!>

// TESTCASE NUMBER: 3
val value_3 = <!ILLEGAL_CONST_EXPRESSION!>.00e-<!>

// TESTCASE NUMBER: 4
val value_4 = <!ILLEGAL_CONST_EXPRESSION!>.000E+<!>

// TESTCASE NUMBER: 5
val value_5 = <!ILLEGAL_CONST_EXPRESSION!>.0e<!>

// TESTCASE NUMBER: 6
val value_6 = <!ILLEGAL_CONST_EXPRESSION!>.00E+<!>

// TESTCASE NUMBER: 7
val value_7 = <!ILLEGAL_CONST_EXPRESSION!>.000e<!>

// TESTCASE NUMBER: 8
val value_8 = <!ILLEGAL_CONST_EXPRESSION!>.0000E<!>

// TESTCASE NUMBER: 9
val value_9 = <!ILLEGAL_CONST_EXPRESSION!>.0e+<!>

// TESTCASE NUMBER: 10
val value_10 = <!ILLEGAL_CONST_EXPRESSION!>.00E-<!>

// TESTCASE NUMBER: 11
val value_11 = <!ILLEGAL_CONST_EXPRESSION!>.000e<!>

// TESTCASE NUMBER: 12
val value_12 = <!ILLEGAL_CONST_EXPRESSION!>.0000e<!>

// TESTCASE NUMBER: 13
val value_13 = <!ILLEGAL_CONST_EXPRESSION!>.0E-<!>

// TESTCASE NUMBER: 14
val value_14 = <!ILLEGAL_CONST_EXPRESSION!>.888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888e+<!>

// TESTCASE NUMBER: 15
val value_15 = <!ILLEGAL_CONST_EXPRESSION!>.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e<!>

// TESTCASE NUMBER: 16
val value_16 = <!ILLEGAL_CONST_EXPRESSION!>.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e-<!>

// TESTCASE NUMBER: 17
val value_17 = <!ILLEGAL_CONST_EXPRESSION!>.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e+<!>
