// TESTCASE NUMBER: 1
val value_1 = <!ILLEGAL_CONST_EXPRESSION!>.0EF<!>

// TESTCASE NUMBER: 2
val value_2 = <!ILLEGAL_CONST_EXPRESSION!>.0ef<!>

// TESTCASE NUMBER: 3
val value_3 = <!ILLEGAL_CONST_EXPRESSION!>.00e-f<!>

// TESTCASE NUMBER: 4
val value_4 = <!ILLEGAL_CONST_EXPRESSION!>.000E+F<!>

// TESTCASE NUMBER: 5
val value_5 = <!ILLEGAL_CONST_EXPRESSION!>.0eF<!>

// TESTCASE NUMBER: 6
val value_6 = <!ILLEGAL_CONST_EXPRESSION!>.00E+f<!>

// TESTCASE NUMBER: 7
val value_7 = <!ILLEGAL_CONST_EXPRESSION!>.000ef<!>

// TESTCASE NUMBER: 8
val value_8 = <!ILLEGAL_CONST_EXPRESSION!>.0000Ef<!>

// TESTCASE NUMBER: 9
val value_9 = <!ILLEGAL_CONST_EXPRESSION!>.0e+F<!>

// TESTCASE NUMBER: 10
val value_10 = <!ILLEGAL_CONST_EXPRESSION!>.00E-F<!>

// TESTCASE NUMBER: 11
val value_11 = <!ILLEGAL_CONST_EXPRESSION!>.000eF<!>

// TESTCASE NUMBER: 12
val value_12 = <!ILLEGAL_CONST_EXPRESSION!>.0000ef<!>

// TESTCASE NUMBER: 13
val value_13 = <!ILLEGAL_CONST_EXPRESSION!>.0E-F<!>

// TESTCASE NUMBER: 14
val value_14 = <!ILLEGAL_CONST_EXPRESSION!>.888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888888e+f<!>

// TESTCASE NUMBER: 15
val value_15 = <!ILLEGAL_CONST_EXPRESSION!>.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000eF<!>

// TESTCASE NUMBER: 16
val value_16 = <!ILLEGAL_CONST_EXPRESSION!>.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e-f<!>

// TESTCASE NUMBER: 17
val value_17 = <!ILLEGAL_CONST_EXPRESSION!>.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e+F<!>
