// FIR_IDENTICAL
// !DIAGNOSTICS: -UNUSED_PARAMETER

<!CONFLICTING_JVM_DECLARATIONS, CONFLICTING_JVM_DECLARATIONS!>enum class E {
    A;

    <!CONFLICTING_JVM_DECLARATIONS!>fun values(): Array<E> = null!!<!>
    <!CONFLICTING_JVM_DECLARATIONS!>fun valueOf(s: String): E = null!!<!>
}<!>
