// TARGET_BACKEND: JVM_OLD

<!CONFLICTING_JVM_DECLARATIONS!>object O<!> {
    <!CONFLICTING_JVM_DECLARATIONS!>val INSTANCE: O<!> = null!!
}

<!CONFLICTING_JVM_DECLARATIONS!>object O2<!> {
    <!CONFLICTING_JVM_DECLARATIONS!>lateinit var INSTANCE: O2<!>
}