// RUN_PIPELINE_TILL: FRONTEND
// FIR_IDENTICAL
typealias<!SYNTAX!><!>

<!SYNTAX!><!>typealias A1<!SYNTAX!><!>

<!SYNTAX!><!>typealias A2 =

<!SYNTAX!><!>typealias Valid = String

/* GENERATED_FIR_TAGS: typeAliasDeclaration */
