// RUN_PIPELINE_TILL: FRONTEND
// FIR_IDENTICAL
<!WRONG_MODIFIER_CONTAINING_DECLARATION!>inner<!> <!INCOMPATIBLE_MODIFIERS!>data<!> class Outer(val x: Int)

/* GENERATED_FIR_TAGS: classDeclaration, data, inner, primaryConstructor, propertyDeclaration */
