/*
 * KOTLIN DIAGNOSTICS SPEC TEST (POSITIVE)
 *
 * SPEC VERSION: 0.1-draft
 * PLACE: constant-literals, real-literals -> paragraph 4 -> sentence 1
 * NUMBER: 1
 * DESCRIPTION: Real literals with underscores in a whole-number part and a fraction part.
 */

// TESTCASE NUMBER: 1
val value_1 = 0.0_0

// TESTCASE NUMBER: 2
val value_2 = 0.0__0___0

// TESTCASE NUMBER: 3
val value_3 = 0.0_0__0_0

// TESTCASE NUMBER: 4
val value_4 = 0__0.0

// TESTCASE NUMBER: 5
val value_5 = 0_0_0.0______0

// TESTCASE NUMBER: 6
val value_6 = 00_______________00.0_0_0

// TESTCASE NUMBER: 7
val value_7 = 2_2.0_0

// TESTCASE NUMBER: 8
val value_8 = 33__3.00__0

// TESTCASE NUMBER: 9
val value_9 = 4_44____4.00______00

// TESTCASE NUMBER: 10
val value_10 = 5_________555_________5.0

// TESTCASE NUMBER: 11
val value_11 = 666_666.0_____________________________________________________________________________________________________________________0

// TESTCASE NUMBER: 12
val value_12 = 7777777.0_0_0

// TESTCASE NUMBER: 13
val value_13 = 8888888_8.0000

// TESTCASE NUMBER: 14
val value_14 = 9_______9______9_____9____9___9__9_9.0

// TESTCASE NUMBER: 15
val value_15 = 0_0_0_0_0_0_0_0_0_0.1234567890

// TESTCASE NUMBER: 16
val value_16 = 1_2_3_4_5_6_7_8_9.2_3_4_5_6_7_8_9

// TESTCASE NUMBER: 17
val value_17 = 234_5_678.345______________678

// TESTCASE NUMBER: 18
val value_18 = 3_456_7.4567

// TESTCASE NUMBER: 19
val value_19 = 456.5_6

// TESTCASE NUMBER: 20
val value_20 = 5.6_5

// TESTCASE NUMBER: 21
val value_21 = 6_54.7654

// TESTCASE NUMBER: 22
val value_22 = 7_6543.87654_3

// TESTCASE NUMBER: 23
val value_23 = 876543_____________2.9_____________8765432

// TESTCASE NUMBER: 24
val value_24 = 9_____________87654321.098765432_____________1

// TESTCASE NUMBER: 25
val value_25 = 000000000000000000000000000000000000000000000000000000000000000000000000000000000000000___0.000000000000000000000000000000000000000000000000000000000000000000000000000000000000000_0

// TESTCASE NUMBER: 26
val value_26 = <!FLOAT_LITERAL_CONFORMS_ZERO!>0_000000000000000000000000000000000000000000000000000000000000000000000000_000000000000000000000000000000000000000000000000000000000000000000000000_000000000000000000000000000000000000000000000000000000000000000000000000_000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.0___0000000000000000000000000000000000000000000000000000000000000000000000000000000___0000000000000000000000000000000000000000000000000000000000000000000000000000000___0000000000000000000000000000000000000000000000000000000000000000000000000000000___0000000000000000000000000000000000000000000000000000000000000000000000000000000___0000000000000000000000000000000000000000000000000000000000000000000000000000000___0000000000000000000000000000000000000000000000000000000000000000000000000000000000000001<!>

// TESTCASE NUMBER: 27
val value_27 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>9999999999999999999999999999999999999999999_______________99999999999999999999999999999_______________99999999999999999999999999999_______________99999999999999999999999999999_______________999999999999999999999999999999999999999999999_______________99999999999999999999999999999_______________99999999999999999999999999999_______________999999999999999999999999999999999999999999999_______________99999999999999999999999999999_______________99999999999999999999999999999_______________999999999999999999999999999999999999999999999.33333333333333333333333333333333333333333333333_33333333333333333333333333333333333333333<!>
