annotation class A() {
    <!ANNOTATION_CLASS_MEMBER!>constructor(s: Nothing?) {}<!>
    <!ANNOTATION_CLASS_MEMBER!>init {}<!>
    <!ANNOTATION_CLASS_MEMBER!>fun foo() {}<!>
    <!ANNOTATION_CLASS_MEMBER!>val bar: Nothing?<!>
    <!ANNOTATION_CLASS_MEMBER!>val baz get() = Unit<!>
}