interface <!CONSTRUCTOR_IN_INTERFACE!>A(val s: String)<!>

interface <!CONSTRUCTOR_IN_INTERFACE!>B constructor(val s: String)<!>

interface C {
    <!CONSTRUCTOR_IN_INTERFACE!>constructor(val s: String)<!> {}
}