// RUN_PIPELINE_TILL: FRONTEND
val foo = <!UNRESOLVED_REFERENCE!>{ <!UNRESOLVED_REFERENCE!>bar<!> }<!>(

<!SYNTAX!><!SYNTAX!><!>val<!> bar <!SYNTAX!><!SYNTAX!>=<!> <!RECURSION_IN_IMPLICIT_TYPES!>{ <!TYPECHECKER_HAS_RUN_INTO_RECURSIVE_PROBLEM, TYPECHECKER_HAS_RUN_INTO_RECURSIVE_PROBLEM!>foo<!> }<!><!>
