/*
 * KOTLIN DIAGNOSTICS SPEC TEST (POSITIVE)
 *
 * SECTIONS: constant-literals, real-literals
 * PARAGRAPH: 3
 * SENTENCE: [2] The fraction part may be omitted only together with the decimal point, if the whole-number part and either the exponent part or the type suffix are present.
 * NUMBER: 2
 * DESCRIPTION: Real literals with omitted a fraction part and an exponent mark.
 */

// TESTCASE NUMBER: 1
val value_1 = 0e0

// TESTCASE NUMBER: 2
val value_2 = 00e00

// TESTCASE NUMBER: 3
val value_3 = 000E-10

// TESTCASE NUMBER: 4
val value_4 = 0000e+00000000000

// TESTCASE NUMBER: 5
val value_5 = 00000000000000000000000000000000000000E1

// TESTCASE NUMBER: 6
val value_6 = 1e1

// TESTCASE NUMBER: 7
val value_7 = 22E-1

// TESTCASE NUMBER: 8
val value_8 = 333e-00000000000

// TESTCASE NUMBER: 9
val value_9 = <!FLOAT_LITERAL_CONFORMS_ZERO!>4444E-99999999999999999<!>

// TESTCASE NUMBER: 10
val value_10 = 55555e10

// TESTCASE NUMBER: 11
val value_11 = 666666E00010

// TESTCASE NUMBER: 12
val value_12 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>7777777e09090909090<!>

// TESTCASE NUMBER: 13
val value_13 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>88888888e1234567890<!>

// TESTCASE NUMBER: 14
val value_14 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>999999999E1234567890<!>

// TESTCASE NUMBER: 15
val value_15 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>123456789e987654321<!>

// TESTCASE NUMBER: 16
val value_16 = 2345678E0

// TESTCASE NUMBER: 17
val value_17 = 34567E+010

// TESTCASE NUMBER: 18
val value_18 = <!FLOAT_LITERAL_CONFORMS_ZERO!>456e-09876543210<!>

// TESTCASE NUMBER: 19
val value_19 = <!FLOAT_LITERAL_CONFORMS_INFINITY!>5e505<!>

// TESTCASE NUMBER: 20
val value_20 = 654e5

// TESTCASE NUMBER: 21
val value_21 = <!FLOAT_LITERAL_CONFORMS_ZERO!>76543E-91823<!>

// TESTCASE NUMBER: 22
val value_22 = 8765432e+90

// TESTCASE NUMBER: 23
val value_23 = 987654321e-1
