// RUN_PIPELINE_TILL: FRONTEND
<!WRONG_MODIFIER_TARGET!>enum<!> interface Some {
    // Enum part
    <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>D<!>;

    // Interface like part
    fun test()
    val foo: Int
}