// FIR_IDENTICAL

interface <!NAME_CONTAINS_ILLEGAL_CHARS!>Iğ”¹<!>

class <!NAME_CONTAINS_ILLEGAL_CHARS!>Cğ”¹<!> : Iğ”¹

<!NAME_CONTAINS_ILLEGAL_CHARS!>fun fğ”¹()<!> = 1

<!NAME_CONTAINS_ILLEGAL_CHARS!>val vğ”¹<!> = 1

typealias ğ”¹ = Boolean
