// Note: JsAllowInvalidCharsIdentifiersEscaping

interface <!NAME_CONTAINS_ILLEGAL_CHARS!>I𝔹<!>

class <!NAME_CONTAINS_ILLEGAL_CHARS!>C𝔹<!> : I𝔹

<!NAME_CONTAINS_ILLEGAL_CHARS!>fun f𝔹()<!> = 1

<!NAME_CONTAINS_ILLEGAL_CHARS!>val v𝔹<!> = 1

typealias 𝔹 = Boolean
