/*
 * KOTLIN CODEGEN BOX SPEC TEST (<!TEST_TYPE!>)
 *
 * SPEC VERSION: 0.1-draft
 * PLACE: <!SECTIONS!> -> paragraph <!PARAGRAPH_NUMBER!> -> sentence <!SENTENCE_NUMBER!>
 * NUMBER: <!TEST_NUMBER!>
 * DESCRIPTION: <!TEST_DESCRIPTION!>
 * NOTE: this test data is generated by FeatureInteractionTestDataGenerator. DO NOT MODIFY CODE MANUALLY!
 */