/*
 * KOTLIN CODEGEN BOX SPEC TEST (<!TEST_TYPE!>)
 *
 * SECTIONS: <!SECTIONS!>
 * PARAGRAPH: <!PARAGRAPH_NUMBER!>
 * SENTENCE: [<!SENTENCE_NUMBER!>] <!SENTENCE!>
 * NUMBER: <!TEST_NUMBER!>
 * DESCRIPTION: <!TEST_DESCRIPTION!>
 * NOTE: this test data is generated by FeatureInteractionTestDataGenerator. DO NOT MODIFY CODE MANUALLY!
 */