/*
 KOTLIN NOT LINKED CODEGEN BOX SPEC TEST (<!TEST_TYPE!>)

 SECTION: <!SECTION!>
 CATEGORIES: <!CATEGORIES!>
 NUMBER: <!TEST_NUMBER!>
 DESCRIPTION: <!TEST_DESCRIPTION!>
 NOTE: this test data is generated by FeatureInteractionTestDataGenerator. DO NOT MODIFY CODE MANUALLY!
 */