// RUN_PIPELINE_TILL: FRONTEND
// ISSUE: KT-65022

val d <!DELEGATION_OPERATOR_MISSING!>by<!> <!UNSUPPORTED!>[]<!>

/* GENERATED_FIR_TAGS: collectionLiteral, nullableType, propertyDeclaration, propertyDelegate, starProjection */
