// RUN_PIPELINE_TILL: FRONTEND
// ISSUE: KT-65022

val d <!DELEGATE_SPECIAL_FUNCTION_MISSING!>by<!> <!UNSUPPORTED_ARRAY_LITERAL_OUTSIDE_OF_ANNOTATION_ERROR!>[]<!>

/* GENERATED_FIR_TAGS: collectionLiteral, nullableType, propertyDeclaration, propertyDelegate, starProjection */
