// RUN_PIPELINE_TILL: FRONTEND
import <!UNRESOLVED_IMPORT!>incorrect<!>.directory.Your

typealias My = <!UNRESOLVED_REFERENCE!>incorrect<!>.directory.My

typealias Your = <!RECURSIVE_TYPEALIAS_EXPANSION!>Your<!>
