// RUN_PIPELINE_TILL: FRONTEND
// ISSUE: KT-76150
// LANGUAGE: +CollectionLiterals

fun test() {
    <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[]<!>]<!>
    <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[<!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>["2"]<!>]<!>
    <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[{}, <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[]<!>]<!>
    <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[::test, <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[2]<!>]<!>
    <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[42, <!UNSUPPORTED_COLLECTION_LITERAL_TYPE!>[]<!>]<!>
}

/* GENERATED_FIR_TAGS: callableReference, collectionLiteral, functionDeclaration, integerLiteral, lambdaLiteral,
stringLiteral */
