// !WITH_NEW_INFERENCE
fun import() {
    <!FUNCTION_CALL_EXPECTED!>import<!> <!UNRESOLVED_REFERENCE!>a<!><!SYNTAX!>.<!><!DEBUG_INFO_MISSING_UNRESOLVED!>*<!><!SYNTAX!><!>
}

fun composite() {
    val <!UNUSED_VARIABLE!>s<!> = 13+<!SYNTAX!>~<!><!DEBUG_INFO_MISSING_UNRESOLVED!>/<!>12
}

fun html() {
    <!NI;SYNTAX!><<!><!NI;FUNCTION_CALL_EXPECTED, OI;DEBUG_INFO_MISSING_UNRESOLVED!>html<!><!NI;UNRESOLVED_REFERENCE!>><!><!NI;SYNTAX!><<!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!>/<!><!DEBUG_INFO_MISSING_UNRESOLVED!>html<!><!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>><!><!NI;SYNTAX!><!>
}

fun html1() {
    <!NI;SYNTAX!><<!><!NI;FUNCTION_CALL_EXPECTED, OI;DEBUG_INFO_MISSING_UNRESOLVED!>html<!><!NI;UNRESOLVED_REFERENCE!>><!><!NI;SYNTAX!><<!><!NI;DEBUG_INFO_MISSING_UNRESOLVED!>/<!><!DEBUG_INFO_MISSING_UNRESOLVED!>html<!><!NI;DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>><!><!FUNCTION_CALL_EXPECTED!>html<!>
}