// DIAGNOSTICS: -REDECLARATION

package<!SYNTAX!><!>

<!FUNCTION_DECLARATION_WITH_NO_NAME!>fun ()<!> {

}

val<!SYNTAX!><!> : Int = 1


class<!SYNTAX!><!> {

}

interface<!SYNTAX!><!> {

}

object<!SYNTAX!><!> {

}

enum class<!SYNTAX!><!> {}

annotation class<!SYNTAX!><!> {}
