// RUN_PIPELINE_TILL: FRONTEND
// FILE: f1.kt
package test

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>A<!>
class F1

// FILE: f2.kt
package test

class <!PACKAGE_OR_CLASSIFIER_REDECLARATION!>A<!>
class F2