// RUN_PIPELINE_TILL: FRONTEND
enum class <!REDECLARATION, REDECLARATION!>A<!> {
    <!REDECLARATION!>name<!>,
    <!REDECLARATION!>ordinal<!>,
    <!DEPRECATED_DECLARATION_OF_ENUM_ENTRY!>entries<!>,
}
