// KT-9733 No error shown for 2 "main" functions in the same file

<!CONFLICTING_OVERLOADS!>fun main(<!UNUSED_PARAMETER!>args<!>: Array<String>)<!> {}
<!CONFLICTING_OVERLOADS!>fun main(<!UNUSED_PARAMETER!>args<!>: Array<String>)<!> {}