<!REDECLARATION!>val Test1 = null<!>
<!REDECLARATION!>class Test1<!>

<!REDECLARATION!>val Test2 = null<!>
<!REDECLARATION!>interface Test2<!>

<!REDECLARATION!>val Test3 = null<!>
<!REDECLARATION!>object Test3<!>

<!REDECLARATION!>val Test4 = null<!>
<!REDECLARATION!>class Test4<!>
<!REDECLARATION!>interface Test4<!>
<!REDECLARATION!>object Test4<!>