<!WRONG_MODIFIER_TARGET!>external<!> class C
<!WRONG_MODIFIER_TARGET!>external<!> object O
<!WRONG_MODIFIER_TARGET!>external<!> interface I
<!WRONG_MODIFIER_TARGET!>external<!> enum class E
<!WRONG_MODIFIER_TARGET!>external<!> annotation class A
<!WRONG_MODIFIER_TARGET!>external<!> typealias TA = Any
<!WRONG_MODIFIER_TARGET!>external<!> val p: Any
