import kotlin.*
import kotlin.jvm.*
import kotlin.native.concurrent.*
import kotlin.native.*

@<!DEPRECATION_ERROR, TYPEALIAS_EXPANSION_DEPRECATION_ERROR!>SharedImmutable<!>
@<!DEPRECATION_ERROR!>ThreadLocal<!>
val x = 42

@Throws(Exception::class)
fun test() {}
