// RUN_PIPELINE_TILL: BACKEND
// ISSUE: KT-81221 KT-81567
import kotlin.annotation.AnnotationTarget.CLASS
import kotlin.annotation.AnnotationTarget.ANNOTATION_CLASS
import kotlin.annotation.AnnotationTarget.TYPE_PARAMETER
import kotlin.annotation.AnnotationTarget.PROPERTY
import kotlin.annotation.AnnotationTarget.FIELD
import kotlin.annotation.AnnotationTarget.LOCAL_VARIABLE
import kotlin.annotation.AnnotationTarget.VALUE_PARAMETER
import kotlin.annotation.AnnotationTarget.CONSTRUCTOR
import kotlin.annotation.AnnotationTarget.FUNCTION
import kotlin.annotation.AnnotationTarget.PROPERTY_GETTER
import kotlin.annotation.AnnotationTarget.PROPERTY_SETTER
import kotlin.annotation.AnnotationTarget.TYPE
import kotlin.annotation.AnnotationTarget.EXPRESSION
import kotlin.annotation.AnnotationTarget.FILE
import kotlin.annotation.AnnotationTarget.TYPEALIAS

@DslMarker
public annotation class A0

@DslMarker
@Retention(AnnotationRetention.SOURCE)
@Target(
    CLASS,
    ANNOTATION_CLASS,
    TYPE_PARAMETER,
    PROPERTY,
    FIELD,
    LOCAL_VARIABLE,
    VALUE_PARAMETER,
    CONSTRUCTOR,
    FUNCTION,
    PROPERTY_GETTER,
    PROPERTY_SETTER,
    TYPE,
    EXPRESSION,
    FILE,
    TYPEALIAS,
)
public annotation class A1

@DslMarker
@Retention(AnnotationRetention.SOURCE)
@Target(
    AnnotationTarget.CLASS,
    AnnotationTarget.ANNOTATION_CLASS,
    AnnotationTarget.TYPE_PARAMETER,
    AnnotationTarget.PROPERTY,
    AnnotationTarget.FIELD,
    AnnotationTarget.LOCAL_VARIABLE,
    AnnotationTarget.VALUE_PARAMETER,
    AnnotationTarget.CONSTRUCTOR,
    AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER,
    AnnotationTarget.PROPERTY_SETTER,
    AnnotationTarget.TYPE,
    AnnotationTarget.EXPRESSION,
    AnnotationTarget.FILE,
    AnnotationTarget.TYPEALIAS,
)
public annotation class A2

<!DSL_MARKER_APPLIED_TO_WRONG_TARGET("A0; top level function")!>@A0<!> <!DSL_MARKER_APPLIED_TO_WRONG_TARGET!>@A1<!> <!DSL_MARKER_APPLIED_TO_WRONG_TARGET!>@A2<!>
fun foo() {}

<!DSL_MARKER_APPLIED_TO_WRONG_TARGET("A0; top level property with backing field")!>@A0<!> <!DSL_MARKER_APPLIED_TO_WRONG_TARGET!>@A1<!> <!DSL_MARKER_APPLIED_TO_WRONG_TARGET!>@A2<!>
val bar = ""

@DslMarker
@Retention(AnnotationRetention.SOURCE)
@Target(
    CLASS,
    TYPE,
)
public annotation class A3

@DslMarker
@Retention(AnnotationRetention.SOURCE)
@Target(
    AnnotationTarget.CLASS,
    AnnotationTarget.TYPE,
)
public annotation class A4

/* GENERATED_FIR_TAGS: annotationDeclaration */
