// RUN_PIPELINE_TILL: FRONTEND
// FIR_IDENTICAL
class A : (<!UNSUPPORTED!>categoryName<!>: <!SYNTAX!><!>)<!SYNTAX!><!> <!SYNTAX!><!>{<!SYNTAX!><!>

/* GENERATED_FIR_TAGS: classDeclaration, functionalType */
