fun foo(): List<Int> = for (<!CANNOT_INFER_PARAMETER_TYPE!>i<!> in <!COMPREHENSION_FLAT_MAP_MISSING!>10<!>, <!CANNOT_INFER_PARAMETER_TYPE!>j<!> in <!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>i<!>) yield <!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>i<!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>*<!><!DEBUG_INFO_ELEMENT_WITH_ERROR_TYPE!>i<!>