// RUN_PIPELINE_TILL: FRONTEND
// FIR_IDENTICAL
// WITH_STDLIB

@JvmInline
value class R(<!SYNTAX!>throw<!> <!SYNTAX!>[<!><!SYNTAX!><!>FileSystemException<!SYNTAX!><!>,  FileSystemException<!SYNTAX!>]<!>)
