// LANGUAGE: +ContextParameters
// IGNORE_ERRORS
class A
class B

<!CONTEXT_PARAMETERS_UNSUPPORTED!>context(a: <!DEBUG_INFO_MISSING_UNRESOLVED!>A<!>)<!>
fun foo(){}

fun foo(<!UNUSED_PARAMETER!>a<!>: A){}

<!CONTEXT_PARAMETERS_UNSUPPORTED!>context(a: <!DEBUG_INFO_MISSING_UNRESOLVED!>A<!>, b: <!DEBUG_INFO_MISSING_UNRESOLVED!>B<!>)<!>
fun bar(){}

<!CONTEXT_PARAMETERS_UNSUPPORTED!>context(a: <!DEBUG_INFO_MISSING_UNRESOLVED!>A<!>)<!>
fun bar(<!UNUSED_PARAMETER!>b<!>: B){}

